features/26-mfa/implementation.md

```diff
@@ -0,0 +1,75 @@
+# Database Tables
+
+## _auth_authenticator
+
+|Column|Type|
+|------|----|
+|id|TEXT|
+|type|TEXT; `totp`, `oob`, `recovery_code`, `bearer_token`|
+|user_id|TEXT; foreign key to `_core_user.id`|
+
+## _auth_authenticator_totp
+
+|Column|Type|
+|------|----|
+|id|TEXT; foreign key to `_auth_authenticator.id`|
+|secret|TEXT|
+|activated|BOOLEAN|
+|created_at|timestamp|
+|activated_at|timestamp|
+|display_name|TEXT|
+
+- `activated` default to `false`.
+- Only `activated = true` is considered an authenticator.
+
+## _auth_authenticator_oob
+
+|Column|Type|
+|------|----|
+|id|TEXT; foreign key to `_auth_authenticator.id`|
+|channel|TEXT; `sms` or `email`|
+|phone|TEXT|
+|email|TEXT|
+|activated|BOOLEAN|
+|created_at|timestamp|
+|activated_at|timestamp|
+
+- `activated` default to `false`.
+- Only `activated = true` is considered an authenticator.
+
+## _auth_authenticator_oob_code
+
+|Column|Type|
+|------|----|
+|id|TEXT|
+|user_id|TEXT; foreign key to `_core_user.id`|
+|code|TEXT|
+|created_at|timestamp|
+|expire_at|timestamp|
+
+- A OOB code is generated when trigger is called.
```

Comment:
Is that a non-expired OOB code would remains valid even if the authenticator that generated it is removed?

Reply:
My original thought is the OOB code is shared across all OOB authenticators. So trigger via different channels will deliver the same OOB code. If we want to have the behavior you mentioned, OOB code should be authenticator specific.

Reply:
So a user can use a SMS delivered code to authenticate as Email OOB authenticator? It seems strange that the `amr` claim does not necessarily reflect the actual delivery channel.

Reply:
`amr` does reflect the channel.